spec/workspace-spec.coffee

```diff
@@ -941,6 +941,113 @@ describe "Workspace", ->
                 .then ->
                   expect(resultPaths).toEqual([file2])
 
+        describe "when a custom directory searcher is registered", ->
+          it "can override the DefaultDirectorySearcher on a per-directory basis", ->
+            foreignFilePath = 'ssh://foreign-directory:8080/hello.txt'
+            numPathsSearchedInDir2 = 1
+            numPathsToPretendToSearchInCustomDirectorySearcher = 10
+            class CustomDirectorySearch
+              constructor: (delegate) ->
+                @promise = Promise.resolve()
+                searchResult1 =
+                  filePath: foreignFilePath,
+                  matches: [
+                    {
+                      lineText: 'Hello world',
+                      lineTextOffset: 0,
+                      matchText: 'Hello',
+                      range: [[0, 0], [0, 5]],
+                    },
+                  ]
+                delegate.didMatch(searchResult1)
+                delegate.didSearchPaths(numPathsToPretendToSearchInCustomDirectorySearcher)
+              then: (args...) ->
+                @promise.then.apply(@promise, args)
+              cancel: ->
+
+            class CustomDirectorySearcher
+              canSearchDirectory: (directory) -> directory.getPath() is dir1
+              search: (directory, delegate, options) ->
+                new CustomDirectorySearch(delegate)
+
+            atom.packages.serviceHub.provide(
+              "atom.directory-searcher", "0.1.0", new CustomDirectorySearcher())
```

Comment:
FYI, this is made difficult by:

```coffee
            waitsForPromise ->
              atom.workspace.scan /aaaa/, {onPathsSearched}, ({filePath}) ->
                resultPaths.push(filePath)
```

Because the `Promise` returned by `atom.workspace.scan` needs `fakeSearch` to do work in order for it to ever resolve.